### ISSUE TITLE
VAE model causes "tensorflow:Gradients do not exist for variables" error

### ISSUE THREAD

**Issue by markalavin** [tensorflow/tensorflow#51247]
I am constructing a Variational Auto-Encoder (VAE) model in Tensorflow 2.15.0, but when I go to fit it, I get the "tensorflow:Gradients do not exist for variables" for the variables in the decoder part of the VAE.

Here is the code for the encoder:
```
    encoder = Sequential( [
                           
        Conv2D( filters = 32, kernel_size = ( 4, 4 ), strides = ( 2, 2 ), padding = "same", activation = "relu",
                input_shape = INPUT_IMAGE_SHAPE, name = "encoder_conv2d_0" ),
        BatchNormalization( name = "encoder_batchnorm_0"),
        Conv2D( filters = 64, kernel_size = ( 4, 4 ), strides = ( 2, 2 ), padding = "same", activation = "relu", name = "encoder_conv2d_1" ),
        BatchNormalization(  name = "encoder_batchnorm_1" ),
        Flatten( name = "encoder_flatten"),
        Dense( tfpl.MultivariateNormalTriL.params_size( latent_dim ), name = "encoder_dense" ),
        tfpl.MultivariateNormalTriL( latent_dim, activity_regularizer = kl_regularizer, dtype = tf.float64, name = "encoder_outdist" )


    ], name = "encoder" )
```
and for the decoder:
```
    decoder = Sequential( [
        
        Dense( units = 9 * 9 * 64, activation = "relu", input_shape = ( latent_dim, ), name = "decoder_dense" ),
        Reshape( ( 9, 9, 64 ) , name = "decoder.reshape" ),
        UpSampling2D( size = ( 2, 2 ), name = "decoder_upsampl_0" ),      # ### 18 x 18 x 32
        Conv2D( filters = 128, kernel_size = ( 3, 3 ), padding = "same", activation = "relu", name = "decoder_conv2D_0" ),
        UpSampling2D( size = ( 2, 2 ), name = "decoder_upsampl_1" ),  # ### 36 x 36 x 16
        Conv2D( filters = 128, kernel_size = ( 3, 3 ), padding = "same", activation = "relu", name = "decoder_conv2D_1" ),
        Conv2D( filters = 3, kernel_size = ( 3, 3 ), padding = "same", name = "decoder_conv2D_2" ),
        Flatten( name = "decoder_flatten" ),
        tfpl.IndependentBernoulli( event_shape = INPUT_IMAGE_SHAPE, name = "decoder_outdist" )
        
    ], name = "decoder" )
```
Here are the model summaries for the encoder and decoder:
```
Model: "encoder"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
encoder_conv2d_0 (Conv2D)    (None, 18, 18, 32)        1568      
_________________________________________________________________
encoder_batchnorm_0 (BatchNo (None, 18, 18, 32)        128       
_________________________________________________________________
encoder_conv2d_1 (Conv2D)    (None, 9, 9, 64)          32832     
_________________________________________________________________
encoder_batchnorm_1 (BatchNo (None, 9, 9, 64)          256       
_________________________________________________________________
encoder_flatten (Flatten)    (None, 5184)              0         
_________________________________________________________________
encoder_dense (Dense)        (None, 5)                 25925     
_________________________________________________________________
encoder_outdist (Multivariat multiple                  0         
=================================================================
Total params: 60,709
Trainable params: 60,517
Non-trainable params: 192
_________________________________________________________________

 None
Model: "decoder"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
decoder_dense (Dense)        (None, 5184)              15552     
_________________________________________________________________
decoder.reshape (Reshape)    (None, 9, 9, 64)          0         
_________________________________________________________________
decoder_upsampl_0 (UpSamplin (None, 18, 18, 64)        0         
_________________________________________________________________
decoder_conv2D_0 (Conv2D)    (None, 18, 18, 128)       73856     
_________________________________________________________________
decoder_upsampl_1 (UpSamplin (None, 36, 36, 128)       0         
_________________________________________________________________
decoder_conv2D_1 (Conv2D)    (None, 36, 36, 128)       147584    
_________________________________________________________________
decoder_conv2D_2 (Conv2D)    (None, 36, 36, 3)         3459      
_________________________________________________________________
decoder_flatten (Flatten)    (None, 3888)              0         
_________________________________________________________________
decoder_outdist (Independent multiple                  0         
=================================================================
Total params: 240,451
Trainable params: 240,451
Non-trainable params: 0
```
Here is the probabilistic loss function:
```
@tf.function
def reconstruction_loss(batch_of_images, decoding_dist):
    """
    This function should compute and return the average expected reconstruction loss,
    as defined above.
    The function takes batch_of_images (Tensor containing a batch of input images to
    the encoder) and decoding_dist (output distribution of decoder after passing the 
    image batch through the encoder and decoder) as arguments.
    The function should return the scalar average expected reconstruction loss.
    """
    batch_loss = decoding_dist.log_prob( batch_of_images )
    loss = - tf.math.reduce_sum( batch_loss ) / batch_of_images.shape[ 0 ]
    return loss
```
Here is how the encoder and decoder are connected for model-fitting:
```
vae = Model(inputs=encoder.inputs, outputs=decoder(encoder.outputs[ 0 ]) )

optimizer = tf.keras.optimizers.Adam(learning_rate=0.0005)
vae.compile(optimizer=optimizer, loss=reconstruction_loss)
```
and here is how the fitting is done:
```
history = vae.fit( x = train_dataset, validation_data = test_dataset, epochs = 50, verbose = True )
```
Finally, here are the error messages I'm getting from the fitting step:
```
Epoch 1/50
WARNING:tensorflow:Gradients do not exist for variables ['decoder_dense/kernel:0', 'decoder_dense/bias:0', 'decoder_conv2D_0/kernel:0', 'decoder_conv2D_0/bias:0', 'decoder_conv2D_1/kernel:0', 'decoder_conv2D_1/bias:0', 'decoder_conv2D_2/kernel:0', 'decoder_conv2D_2/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['decoder_dense/kernel:0', 'decoder_dense/bias:0', 'decoder_conv2D_0/kernel:0', 'decoder_conv2D_0/bias:0', 'decoder_conv2D_1/kernel:0', 'decoder_conv2D_1/bias:0', 'decoder_conv2D_2/kernel:0', 'decoder_conv2D_2/bias:0'] when minimizing the loss.
40/40 [==============================] - 4s 29ms/step - loss: 0.1177 - val_loss: 0.0263
Epoch 2/50
40/40 [==============================] - 0s 10ms/step - loss: 0.0406 - val_loss: 0.0381
...
```

**Comment by sanatmpa1** [comment]
@markalavin,

Can you please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new/choose) with corresponding details and also please provide a simple reproducible standalone code to expedite the troubleshooting? Thanks!

**Comment by markalavin** [comment]
@sanatmpa1, I am working on creating a Jupyter workspace that reproduces the problem.   I will post here when it's ready to try.

In the meantime, can you offer any advice about what could contribute to the "Gradients do not exist for variables"?  Thanks.

**Comment by markalavin** [comment]
@sanatmpa1 , here is a small test case that exhibits the problem of missing gradients:
```
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
tfb = tfp.bijectors
tfpl = tfp.layers

import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

import math
from math import sin, cos
import traceback

from sklearn.model_selection import train_test_split

from tensorflow.keras import Sequential, Model
from tensorflow.keras.layers import (Dense, Flatten, Reshape, Concatenate, Conv2D, 
                                     UpSampling2D, BatchNormalization)

import sys

print( "tf.version", tf.__version__, "tfp.version", tfp.__version__ )



image_array = np.random.uniform( 0, 1, ( 100, 36, 36, 3 ) ).astype( np.float64 ) / 255.0
train_dataset = tf.data.Dataset.from_tensor_slices( image_array[ 0 : 80 ] )
test_dataset = tf.data.Dataset.from_tensor_slices( image_array[ 80 : 100 ] )
train_dataset = train_dataset.batch( 20 )
test_dataset = test_dataset.batch( 20 )

def reconstruction_loss(batch_of_images, decoding_dist):
    return - tf.reduce_mean( decoding_dist.log_prob( batch_of_images ) )

prior_distribution = tfd.MultivariateNormalDiag( loc = np.zeros( 2 ), scale_diag = np.ones( 2 ) )

kl_regularizer = tfpl.KLDivergenceRegularizer(
    distribution_b = prior_distribution,
    use_exact_kl = False,
    test_points_fn = lambda t : t.sample( 3 ),
    test_points_reduce_axis = None,
    weight = 1.0
)

encoder = tf.keras.Sequential( [
    Conv2D( 3, 3, padding = "same", input_shape = ( 36, 36, 3 ) ),
    Flatten(),
    Dense( tfpl.MultivariateNormalTriL.params_size( 2 ), name = "encoder_dense" ),
    tfpl.MultivariateNormalTriL( 2, activity_regularizer = kl_regularizer, dtype = tf.float64, name = "encoder_outdist" )
])

decoder = tf.keras.Sequential( [
    Dense( 36 * 36 * 3, input_shape = ( 2, ), name = "decoder_dense" ),
    Reshape( ( 36, 36, 3 ), name = "decoder_reshape" ),
    Conv2D( 3, 3, padding = "same", name = "decoder_conv2d" ),
])
vae = tf.keras.Model( inputs=encoder.inputs, outputs=decoder( encoder.outputs ) )
vae.compile( optimizer = "adam", loss = reconstruction_loss )
print( "vae.summary()", vae.summary() )
vae.fit( train_dataset, validation_data = test_dataset, epochs = 5 )
```

**Comment by markalavin** [comment]
Well, it turns out that the problem can be fixed by changing the following two lines in the preceding example code:
```
train_dataset = tf.data.Dataset.from_tensor_slices( ( image_array[ 0 : 80 ], image_array[ 0 : 80 ] ) )
test_dataset = tf.data.Dataset.from_tensor_slices( ( image_array[ 80 : 100 ], image_array[ 80 : 100 ] ) )
```
This seems rather strange to me, since it implies that the input image is being compared with the output image, whereas in
fact the loss function evaluates the probability of getting the output image from the output distribution.   Nevertheless, I'll take it, since it also produces correct-looking results.
